### PR TITLE
[Doc] Repin ML samples to base ray image with runtime-installed deps

### LINKED ISSUE #5123: [Bug] ray-job.batch-inference.yaml pins a rayproject/ray-ml tag that returns 404
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

Others

### What happened + What you expected to happen

`ray-operator/config/samples/ray-job.batch-inference.yaml` pins a container image that doesn't exist, so the sample can't run as written. Line 16:

```yaml
image: rayproject/ray-ml:2.46.0.0e19ea-py39-gpu
```

That tag returns 404 from Docker Hub:

```console
$ curl -s -o /dev/null -w "%{http_code}\n" \
    https://hub.docker.com/v2/repositories/rayproject/ray-ml/tags/2.46.0.0e19ea-py39-gpu
404
```

Applying the sample leaves the head pod in `ImagePullBackOff`, so the RayJob never starts.

Present on both `master` and `v1.6.0`.

The underlying cause is that `rayproject/ray-ml` is no longer published for public consumption. Versioned tags stop at `2.30.0` — `2.31.0` through `2.56.1` all return 404 — and Ray pushed a family of 16 `2.46.0.deprecated*` tags on 2025-05-07. So this isn't a one-off typo in a build hash; there is no working `ray-ml` tag at 2.46.0 to correct it to.

Picking a replacement needs a decision rather than a find-and-replace, because the sample's workload genuinely needs ML libraries that the base `rayproject/ray` images don't ship. The mounted `sample_code.py` imports `transformers` and `PIL` and runs a Hugging Face `image-classification` pipeline on GPU. Two options I can see:

1. Pin `rayproject/ray:<current>-gpu` and add a `pip install transformers pillow torch` step, either in the RayJob's `runtimeEnvYAML` or as a `RUN` layer in a small custom image.
2. Build and publish a KubeRay-side sample image that carries the dependencies, and pin that.

Happy to open a PR for whichever direction maintainers prefer.

### Downstream: the Ray docs quote this file, and quote it inconsistently

`ray-project/ray`'s [`rayjob-batch-inference-example.md`](https://github.com/ray-project/ray/blob/master/doc/source/cluster/kubernetes/examples/rayjob-batch-inference-example.md) tells the reader to `curl` this sample from `v1.6.0` and then quotes the head-node spec back to them — but shows `rayproject/ray-ml:2.6.3-gpu`, not the tag actually in the file. So a reader sees one image in the tutorial prose and pulls another. (`ray-ml:2.6.3-gpu` does still resolve, which makes the mismatch easy to miss.)

Both halves need fixing, but this one is the upstream half: correcting the Ray doc's quote first would just make it accurately quote a broken tag. Once this sample lands on a working image I'll sync the Ray-side snippet to match.

### Reproduction script

```bash
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-job.batch-inference.yaml
kubectl get pods -w   # head pod enters ImagePullBackOff
```

Tag existence alone:

```bash
curl -s -o /dev/null -w "%{http_code}\n" \
  https://hub.docker.com/v2/repositories/rayproject/ray-ml/tags/2.46.0.0e19ea-py39-gpu
```

### Anything else

Found while auditing container image pins across the Ray documentation. AI assistance (Claude Code) was used to inventory the pins and verify each tag against Docker Hub; I reviewed and confirmed the findings above.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by dstrodtman**
## Why are these changes needed?

The `rayproject/ray-ml:2.46.0.0e19ea-*` tag family was deleted from Docker Hub, so five samples land in `ImagePullBackOff` as written:

```console
$ curl -s -o /dev/null -w "%{http_code}\n" \
    https://hub.docker.com/v2/repositories/rayproject/ray-ml/tags/2.46.0.0e19ea-py39-gpu
404
```

This takes the **option 1** direction agreed in #5123: rather than repin to the surviving `2.49.2.deprecated-*` `ray-ml` tags — an unmaintained registry that Ray has said stops publishing entirely from 2.50.x, "not even with `-deprecated` tags" (ray-project/ray#46378) — each sample moves to the published `rayproject/ray:2.58.0` base image and installs the ML dependencies it actually imports through the runtime environment. As @win5923 put it on the issue, the sample YAMLs exist to help users try KubeRay by following the docs, so there's no need to maintain a KubeRay-specific ML image just to run them.

`py39` base tags no longer publish (`rayproject/ray:2.58.0-py39` is a 404), so the three py39 samples move to `py311`.

## What changed, per sample

The dependency each sample needs comes from what its code actually imports, not from guesswork:

| Sample | Image | Runtime deps added |
| --- | --- | --- |
| `ray-service.mobilenet.yaml` | `ray:2.58.0-py311` (CPU) | `tensorflow==2.15.1` |
| `ray-service.stable-diffusion.yaml` | `ray:2.58.0-py311-gpu` | `torch` |
| `ray-job.batch-inference.yaml` | `ray:2.58.0-py311-gpu` | `numpy`, `pillow`, `torch`, `transformers` (new `runtimeEnvYAML`) |
| `pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml` | head `ray:2.58.0-py311`, worker `ray:2.58.0-py311-gpu` | none — `pip:` already declared them |
| `pytorch-text-classifier/ray-job.pytorch-distributed-training.yaml` | `ray:2.58.0-py311-gpu` | `pytorch-lightning>=2.0` (was `==1.8.5`), `torchmetrics` — see note |

Two notes on the choices:

- **`mobilenet` pins `tensorflow==2.15.1` deliberately.** `mobilenet.py` imports `tensorflow.keras.preprocessing.image`, which Keras 3 removed. TensorFlow ships Keras 3 by default from 2.16, so an unpinned `tensorflow` breaks the sample. 2.15.1 is the last Keras-2 line and publishes a cp311 wheel. This matches the version Ray pinned for this stack.
- **`pytorch-distributed-training` moves to Lightning 2.x and `torchmetrics`.** On the base image the sample's deps resolve to Ray 2.58's stack (torch 2.9), and two entries in the old pin set break against it. `pytorch-lightning==1.8.5` predates torch 2.0 and can't run on torch 2.x. The sample code already uses the Lightning 2.x validation pattern (`on_validation_epoch_end` with manual accumulation), so the pin moves to `pytorch-lightning>=2.0`, and `torch` continues to resolve through Lightning (which now requires `torch>=2.1`). Separately, `datasets.load_metric` was removed in `datasets` 3.0 and is absent from the base image's `datasets`; pinning `datasets` back to a 2.x line to restore it would drag in `dill<0.3.9` and pre-1.x `huggingface-hub`, which conflict with Ray's own pins. The metric is swapped to `torchmetrics.BinaryMatthewsCorrCoef`, which ships with Lightning, aggregates across workers, and needs no runtime metric-script download. This tracks the current Ray docs `lightning_cola_advanced` example, which likewise dropped `load_metric` and moved to `lightning>=2.0`.

## Related issue number

Closes #5123. Supersedes #5129, which repinned the same five files to the `.deprecated` `ray-ml` tags.

**Companion docs PR:** ray-project/ray#65708 syncs the Ray docs pages that quote these manifests (mobilenet, batch-inference, stable-diffusion). It's marked to merge after this one lands on `master`.

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [x] Manual tests

### Manual test instructions

**Every image tag referenced by the five changed files resolves on Docker Hub** (the same loop on `master` returns 404 for the `ray-ml` lines):

```console
$ for img in rayproject/ray:2.58.0-py311 rayproject/ray:2.58.0-py311-gpu; do
    printf "%s %s\n" \
      "$(curl -s -o /dev/null -w '%{http_code}' "https://hub.docker.com/v2/repositories/${img%%:*}/tags/${img##*:}")" "$img"
  done
200 rayproject/ray:2.58.0-py311
200 rayproject/ray:2.58.0-py311-gpu
```

**Not run: full end-to-end execution.** `stable-diffusion` and `batch-inference` need GPUs and `pytorch-image-classifier` needs a GKE GCSFuse service account, none of which I can exercise here, and none of these five samples are in the `test/sampleyaml` CI matrix. A GPU-side confirmation of the conversions is welcome.

**Dependency resolution audited by introspection.** The dependency sets come from each sample's own imports, and each was checked to resolve against Ray 2.58's validated pins (torch 2.9.0, transformers 5.11.0, datasets 5.0.1, numpy 2.2.6, pillow 11.3.0) and the base image's transitive pins (`dill` 0.4.1, `fsspec` 2023.12.1, `huggingface-hub` 1.16.1, `pyarrow` 23.0.1). Four samples resolve cleanly as written. The `pytorch-resnet-image-classifier` sample keeps its existing `pip:` list unchanged; its `datasets` and `transformers` entries are unused by the code (the entrypoint imports `torchvision.datasets`) and pre-existing, and they resolve cleanly now that modern `datasets` caps `dill<0.4.2`. Only `pytorch-text-classifier` needed the Lightning-and-metric change above, because its old pins don't resolve to a working sample on the base image.

### CODE REVIEW

**Comment by dstrodtman** [comment]
Pushed an update addressing both signals on the first run:

- **Cursor Bugbot (Missing Pillow for MobileNet):** good catch. `mobilenet.py` loads images through `keras.preprocessing.image`, which needs PIL at request time, and the base `rayproject/ray` image doesn't ship Pillow the way `ray-ml` did. Added `pillow` to the sample's `runtime_env` pip list.
- **`Build Operator Binaries and Docker Images` failure:** that was an unrelated `go vet` error (`undefined: record` in `rayservice_controller_unit_test.go`), which isn't present on current `master` — this branch was based on an older commit. Rebased onto `master`, which clears it. This PR only changes sample YAMLs.

File: ray-operator/config/samples/pytorch-text-classifier/ray-job.pytorch-distributed-training.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ spec:
             effect: "NoSchedule"
           containers:
           - name: ray-head
-            image: rayproject/ray-ml:2.46.0.0e19ea-py311-gpu
+            image: rayproject/ray:2.58.0-py311-gpu
```
[the remark this answers is not in the document]
Comment (dstrodtman) [reply ray-operator/config/samples/pytorch-text-classifier/ray-job.pytorch-distributed-training.yaml:31]: Both halves confirmed and fixed in e402b6c8.

- **`load_metric`**: correct — it was removed in `datasets` 3.0, so it's gone from the base image's `datasets`. Restoring it via a `datasets<3` pin would pull `dill<0.3.9` and pre-1.x `huggingface-hub`, conflicting with Ray 2.58's own pins, so instead the metric moves to `torchmetrics.BinaryMatthewsCorrCoef` (ships with Lightning, DDP-aware, no runtime metric-script download).
- **torch**: the direction is inverted, though — the sample already uses the Lightning 2.x validation pattern and Ray 2.46's `ray-ml` shipped `torch==2.3.0`, so torch 2.x is what it was validated against, not a regression. The real bug was the `pytorch-lightning==1.8.5` pin, which predates torch 2.0; bumped to `pytorch-lightning>=2.0`.

This mirrors the current Ray docs `lightning_cola_advanced` example.

File: ray-operator/config/samples/pytorch-text-classifier/fine-tune-pytorch-text-classifier.py [hunk 2]
```diff
@@ -46,21 +45,13 @@ def validation_step(self, batch, batch_idx):
...
-
-        # self.metric.compute() returns a dictionary:
-        # e.g. {"matthews_correlation": 0.53}
-        self.log_dict(matthews_correlation, sync_dist=True)
-        self.predictions.clear()
-        self.references.clear()
+        # torchmetrics aggregates state across workers, so compute() returns the
+        # Matthews correlation over the full validation set.
+        self.log("matthews_correlation", self.metric.compute(), sync_dist=True)
+        self.metric.reset()
```
Comment (win5923) [inline ray-operator/config/samples/pytorch-text-classifier/fine-tune-pytorch-text-classifier.py:54]: Have these changes been tested?

**Review by win5923 — APPROVED** [review]

File: ray-operator/config/samples/pytorch-text-classifier/fine-tune-pytorch-text-classifier.py (hunk 2 above)
Comment (dstrodtman) [reply ray-operator/config/samples/pytorch-text-classifier/fine-tune-pytorch-text-classifier.py:54]: Nope. Changes here are introspected from source but not run. Importantly, current examples are likely not runnable (at least not using any maintained Ray image), so this should be at worst net-neutral as far as introducing broken code. The only downside to shipping this would be that code appears to be updated (and therefore recently tested and actively maintained) rather than clearly reading as stale.

Agree that full testing is desired. Right now there's no clear maintenance loop feeding back into keeping these examples up-to-date.

File: ray-operator/config/samples/ray-service.stable-diffusion.yaml [hunk 3]
```diff
@@ -21,7 +21,7 @@ spec:
         spec:
           containers:
           - name: ray-head
-            image: rayproject/ray-ml:2.46.0.0e19ea-py39-gpu
+            image: rayproject/ray:2.58.0-py311-gpu
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.stable-diffusion.yaml:24]: I think some of these examples may not work because `-gpu` images and the old `ray-ml` images have different dependencies. But that's still better than the ImagePullErr :)
